sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala

```diff
@@ -83,32 +83,32 @@ case class ApproximatePercentile(
   }
 
   // Mark as lazy so that accuracyExpression is not evaluated during tree transformation.
-  private lazy val accuracy: Int = accuracyExpression.eval().asInstanceOf[Int]
+  private lazy val accuracy: Long = accuracyExpression.eval().asInstanceOf[Number].longValue
 
-  override def inputTypes: Seq[AbstractDataType] = {
+  def inputTypes: Seq[AbstractDataType] = {
     // Support NumericType, DateType and TimestampType since their internal types are all numeric,
     // and can be easily cast to double for processing.
     Seq(TypeCollection(NumericType, DateType, TimestampType),
-      TypeCollection(DoubleType, ArrayType(DoubleType)), IntegerType)
+      TypeCollection(DoubleType, ArrayType(DoubleType, containsNull = false)), IntegralType)
```

Comment:
hmm, seems like this change and https://github.com/apache/spark/pull/26905/files#diff-7448088085a868386c669c69629c8169R86 can fix the bug already?

Reply:
Do you mean re-enable ImplicitCastInputTypes?

Reply:
this change disable nulls in percentages array, the change in the link fix accuracy overflow, there is still a match error if percentage is null

Reply:
yea re-enable `ImplicitCastInputTypes`.

We can implicitly add a check in `checkIntputTypes` for null percentage.

Reply:
we will support `accuracy` with fractions, but it's trivial

Reply:
Then we can accept all `NumericType`, and in `checkInputDataTypes` we fail if the `accuracy` is actually not integral type.

Reply:
I was wrong about https://github.com/apache/spark/pull/26905/files#r359753430, there is no proper rule for `FractionalType` to `IntergralType` in ImplicitTypeCast. It will fail when doing type checking later if it is not an integral, so it just fine here

Reply:
This is can be verified by function percentile it is `ImplicitCastInputTypes` and its 3rd arg is `IntergralType`

```sql
         > select percentile(1, '0.1', 1.2);
Error in query: cannot resolve 'percentile(1, CAST('0.1' AS DOUBLE), 1.2BD)' due to data type mismatch: argument 3 requires integral type, however, '1.2BD' is of decimal(2,1) type.; line 1 pos 7;
'Aggregate [unresolvedalias(percentile(1, cast(0.1 as double), 1.2, 0, 0), None)]
+- OneRowRelation
```